cylc/flow/platform_lookup.py

```diff
@@ -0,0 +1,166 @@
+        platfrom (str):
+            string representing a platform from the global config.
+
+    Raises:
+        PlatformLookupError:
+            If no matching platform can be a found an error is raised.
+
+    Example:
+        >>> platforms = {
+        ...         'desktop[0-9][0-9]|laptop[0-9][0-9]': {},
+        ...         'sugar': {
+        ...             'login hosts': 'localhost',
+        ...             'batch system': 'slurm'
+        ...         }
+        ... }
+        >>> job = {'batch system': 'slurm'}
+        >>> remote = {'host': 'sugar'}
+        >>> reverse_lookup(platforms, job, remote)
+        'sugar'
+        >>> remote = {}
+        >>> reverse_lookup(platforms, job, remote)
+        'localhost'
+    """
+    # These settings are removed from the incoming dictionaries for special
+    # handling later - we want more than a simple match:
+    #   - In the case of host we also want a regex match to the platform name
+    #   - In the case of batch system we want to match the name of the system
+    #     to a platform when host is localhost.
+    if 'host' in remote.keys():
+        task_host = remote.pop('host')
+    else:
+        task_host = 'localhost'
+    if 'batch system' in job.keys():
+        task_batch_system = job.pop('batch system')
+    else:
+        # Necessary? Perhaps not if batch system default is 'background'
+        task_batch_system = 'background'
+
+    # Riffle through the platforms looking for a match to our task settings.
+    # reverse dict order
+    platforms = {
+        key: platforms[key] for key in reversed(list(platforms.keys()))
+    }
+
+    for platform_name, platform_spec in platforms.items():
+        # Handle all the items requiring an exact match.
+        generic_items_match = True
+        for task_section in [job, remote]:
+            shared_items = set(
+                task_section.keys()).intersection(set(platform_spec.keys()))
+            for shared_item in shared_items:
+                # breakpoint()
+                if platform_spec[shared_item] != task_section[shared_item]:
+                    generic_items_match = False
+
+        # All items other than batch system and host must be an exact match
+        if not generic_items_match:
+            continue
+
+        # We have some special logic to identify whether task host and task
+        # batch system match the platform in question.
+        if task_host == 'localhost':
+            if task_batch_system == 'background':
+                return 'localhost'
+            elif (
+                task_batch_system == platform_spec['batch system'] and
+                'hosts' in platform_spec.keys() and
+                'localhost' in platform_spec['hosts']
+            ):
+                # If we have localhost with a non-background batch system we
+                # use the batch system to give a sensible guess at the platform
+                return platform_name
+
+        else:
+            if (
```

Comment:
I like seeing if statements written out like this, much clearer!

This `if` statement looks the same as the one in the previous branch.

Reply:
> This if statement looks the same as the one in the previous branch.

Which previous branch? They are similar bits of logic, but not quite the same. This section is the reason for the flow chart in the docstring.

Reply:
I think these two branches are identical except in the first `task_host` is hardcoded to `localhost`:

```python
            elif (
                task_batch_system == platform_spec['batch system'] and
                'hosts' in platform_spec.keys() and
                'localhost' in platform_spec['hosts']
            ):
```

```python
        if (
                'hosts' in platform_spec.keys() and
                task_host in platform_spec['hosts'] and
                task_batch_system == platform_spec['batch system']
            ):
```

Reply:
Quite right. I was being a blind beetle on this one.